### PR TITLE
[RLlib] APPO+new-stack (Atari benchmark) - Preparatory PR 01

### PR DESCRIPTION
**Pull request by sven1977**
APPO+new-stack (Atari benchmark) - Preparatory PR 01
* Code cleanups.
* tuned_examples for Atari should all configure their env_config to match the old NoFrameskip + Deterministic (-v4) setup (see here for more details: https://gymnasium.farama.org/environments/atari/#version-history-and-naming-schemes). 

This is a breakdown PR (1st of N) from this one here, which is too large to merge: https://github.com/ray-project/ray/pull/34363

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/learner/learner.py [hunk 1]
```diff
@@ -522,14 +522,14 @@ def compile_results(
 
         # We put the stats for all modules under the ALL_MODULES key. e.g. average of
         # the gradients across all modules will go here.
-        mean_grads = [
-            np.mean(grad)
+        mean_abs_grads = [
```
Comment (sven1977) [inline rllib/core/learner/learner.py:525]: MEAN(ABS(grads)) is a much better metric than MEAN(grads) as the latter is always very close to zero, regardless of the nature of the gradients.

File: rllib/policy/torch_mixins.py [hunk 2]
```diff
@@ -184,17 +184,32 @@ def update_target(self, tau=None):
...
         # If tau == 1.0: Full sync from Q-model to target Q-model.
-        target_state_dict = next(iter(self.target_models.values())).state_dict()
-        model_state_dict = {
-            k: tau * model_state_dict[k] + (1 - tau) * v
-            for k, v in target_state_dict.items()
-        }
 
-        for target in self.target_models.values():
-            target.load_state_dict(model_state_dict)
+        if self.config.get("_enable_rl_module_api", False):
```
Comment (sven1977) [inline rllib/policy/torch_mixins.py:193]: Preparation for the upcoming move of APPO (torch) to the new API stack.

File: rllib/tuned_examples/appo/cartpole-appo-learner.yaml [hunk 3]
```diff
@@ -1,31 +0,0 @@
-cartpole-appo-learner:
```
Comment (sven1977) [inline rllib/tuned_examples/appo/cartpole-appo-learner.yaml:1]: Gave this file a better, more descriptive/accurate name.

File: rllib/utils/actor_manager.py [hunk 4]
```diff
@@ -232,8 +232,8 @@ class _ActorState:
     def __init__(
         self,
         actors: Optional[List[ActorHandle]] = None,
-        max_remote_requests_in_flight_per_actor: Optional[int] = 2,
```
Comment (sven1977) [inline rllib/utils/actor_manager.py:235]: These are not optional. They must be ints. Yes, they have default (int) values, but cannot be None.

File: rllib/core/learner/learner.py (hunk 1 above)
Comment (avnishn) [reply rllib/core/learner/learner.py:525]: what we should actually probably do here is the gradient norms. We can stick with this for now as it is similar, but in the future, norm is probably the way to go.

File: rllib/core/learner/learner.py [hunk 5]
```diff
@@ -957,29 +961,29 @@ def _make_module(self) -> MultiAgentRLModule:
...
         module. If the module_class is a single agent RL module it will be wrapped to a
-        multi-agent RL module. Override this method if there are other things than
-        needs to happen for instantiation of the module.
-
+        multi-agent RL module. Override this method if there are other things that
+        need to happen for instantiation of the module.
 
         Returns:
-            The constructed module.
+            The constructed MultiAgentRLModule.
```
Comment (avnishn) [inline rllib/core/learner/learner.py:968]: nit not necessary to address: A constructed MultiAgentRLModule

File: rllib/core/learner/scaling_config.py [hunk 6]
```diff
@@ -13,7 +14,8 @@ class LearnerGroupScalingConfig:
             training will run on a single CPU.
         num_gpus_per_worker: The number of GPUs to allocate per worker. If
             num_workers=0, any number greater than 0 will run the training on a single
-            GPU. A value of zero will run the training on a single CPU.
+            GPU. A value of zero will run the training on a single CPU. Fractional
```
Comment (avnishn) [inline rllib/core/learner/scaling_config.py:17]: fractional values aren't really supported. Fractional gpu ussage will cause cuda aync errors AFAICT

File: rllib/tuned_examples/appo/pong-appo.yaml [hunk 7]
```diff
@@ -12,10 +12,13 @@ pong-appo:
     config:
         # Works for both torch and tf.
         framework: torch
+        # Make analogous to old v4 + NoFrameskip.
         env_config:
             frameskip: 1  # no frameskip
-        vtrace: True
-        use_kl_loss: False
+            full_action_space: false
+            repeat_action_probability: 0.0  # deterministic
```
Comment (avnishn) [inline rllib/tuned_examples/appo/pong-appo.yaml:19]: hmm what is this. What is the repeat action probability by default and how much does it matter?

File: rllib/tuned_examples/appo/pong-appo.yaml [hunk 8]
```diff
@@ -12,10 +12,13 @@ pong-appo:
...
+        vtrace: true
+        use_kl_loss: false
```
Comment (avnishn) [inline rllib/tuned_examples/appo/pong-appo.yaml:21]: do we need to disable the kl loss? Is this something that we were doing in the old stack?

File: rllib/tuned_examples/appo/pong-appo-w-rl-modules-and-learner.yaml [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            conv_filters: [[16, 4, 2], [32, 4, 2], [256, 11, 1, "valid"]]
+            conv_activation: relu
+            conv_add_final_dense: false
+            conv_flattened_dim: 256
+            use_cnn_heads: true
+
+        # Run with Learner API.
+        _enable_learner_api: true
+        grad_clip_by_global_norm: 10.0
+        # Use a single Learner worker on the GPU.
```
Comment (kouroshHakha) [inline rllib/tuned_examples/appo/pong-appo-w-rl-modules-and-learner.yaml:45]: the comment doesn't match the code.

**Review by avnishn — COMMENTED** [review]
some questions, ty for this cleanup

**Review by kouroshHakha — APPROVED** [review]

File: rllib/tuned_examples/appo/pong-appo.yaml (hunk 7 above)
Comment (sven1977) [reply rllib/tuned_examples/appo/pong-appo.yaml:19]: Repeat action prob by default is 025, meaning that in 25% of all `Env.step(action)` calls, the env doesn't actually apply the given action, but repeats the previous one. This means that by default, Atari envs are stochastic, but our benchmarks always used the deterministic setting.
See here for more details: https://gymnasium.farama.org/environments/atari/#version-history-and-naming-schemes

We used to run against: "PongNoFrameskip-v4", where the v4 indicates repeat_action_probability=0.0 and the NoFrameskip means frameskip=0.

File: rllib/core/learner/scaling_config.py (hunk 6 above)
Comment (sven1977) [reply rllib/core/learner/scaling_config.py:17]: Thanks, will fix. ...

File: rllib/tuned_examples/appo/pong-appo.yaml (hunk 8 above)
Comment (sven1977) [reply rllib/tuned_examples/appo/pong-appo.yaml:21]: I noticed in the new benchmarks that KL-loss might destabilize the run, but this observation was made while several other issues had not been fixed so it could have been a red herring. Either way, we don't currently use kl_loss in our ols-stack benchmark, so I would like to keep it switched off until we have fully investigated a possible benefit of using this term.

File: rllib/core/learner/scaling_config.py (hunk 6 above)
Comment (sven1977) [reply rllib/core/learner/scaling_config.py:17]: done

Comment (sven1977) [reply rllib/core/learner/scaling_config.py:17]: Actually, there is one test, where we do set num_gpus_per_worker to 0.5:

In `test_learner_group.py`:
```
LOCAL_SCALING_CONFIGS = {
    "local-cpu": LearnerGroupScalingConfig(num_workers=0, num_gpus_per_worker=0),
    "local-gpu": LearnerGroupScalingConfig(num_workers=0, num_gpus_per_worker=0.5),
}
```

Comment (sven1977) [reply rllib/core/learner/scaling_config.py:17]: I think that's why I fixed the typehint, but yeah, it makes sense that torch/tf DDP wouldn't like it. I fixed the comment and said it's not allowed due to xyz.

File: rllib/tuned_examples/dqn/atari-dist-dqn.yaml [hunk 10]
```diff
@@ -7,8 +7,11 @@ atari-dist-dqn:
             - ALE/SpaceInvaders-v5
     run: DQN
     config:
+        # Make analogous to old v4 + NoFrameskip.
         env_config:
-            frameskip: 1  # no frameskip
+            frameskip: 1
+            full_action_space: false
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/tuned_examples/dqn/atari-dist-dqn.yaml:13]: See here: https://gymnasium.farama.org/environments/atari/#action-space
The default is `False` anyways, so no need to really set this here, but for explicitness, I added this as well.

File: rllib/tuned_examples/appo/cartpole-appo-w-rl-modules-and-learner.yaml [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+        enable_connectors: true
+
+        _enable_learner_api: true
+        train_batch_size: 1000
+        lr: 0.001
+        num_learner_workers: 2
+        num_gpus_per_learner_worker: 0
+        num_cpus_per_learner_worker: 1
+
+        _enable_rl_module_api: true
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/tuned_examples/appo/cartpole-appo-w-rl-modules-and-learner.yaml:25]: done
